src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java

```diff
@@ -175,18 +178,57 @@ public void setCacheCapacityReached(Boolean value) {
      * Loads hnsw index to memory. Registers the location of the serialized graph with ResourceWatcher.
      *
      * @param indexPathUrl path for serialized hnsw graph
-     * @param algoParams hnsw algoparams
+     * @param algoParams hnsw algorithm parameters
      * @return KNNIndex holding the heap pointer of the loaded graph
      * @throws Exception Exception could occur when registering the index path
      * to Resource watcher or if the JNI call throws
      */
-    public KNNIndex loadIndex(String indexPathUrl, final String[] algoParams) throws Exception {
+    public KNNIndexCacheEntry loadIndex(String indexPathUrl, final String[] algoParams) throws Exception {
         if(Strings.isNullOrEmpty(indexPathUrl))
             throw new IllegalStateException("indexPath is null while performing load index");
         logger.debug("Loading index on cache miss .. {}", indexPathUrl);
         Path indexPath = Paths.get(indexPathUrl);
-        knnIndexFileListener.register(indexPath);
-        return KNNIndex.loadIndex(indexPathUrl, algoParams);
+        FileWatcher fileWatcher = new FileWatcher(indexPath);
+        fileWatcher.addListener(KNN_INDEX_FILE_DELETED_LISTENER);
+
+        // Calling init() on the FileWatcher will bootstrap initial state that indicates whether or not the file
+        // is present. If it is not present at time of init(), then KNNIndex.loadIndex will fail and we won't cache
+        // the entry
+        fileWatcher.init();
+
+        final KNNIndex knnIndex = KNNIndex.loadIndex(indexPathUrl, algoParams);
```

Comment:
I think we should first do  ```resourceWatcherService.add``` and then load the index.
- ```resourceWatcherService.add``` seems to already call ```init()``` on the filewatcher.  So no explicit call to init()
-  this will also address the case of  index not loaded if file is only present

Reply:
Can you clarify what you mean by the second case?

For the first case, the reason why I'm calling ```loadIndex``` in between ```FileWatcher.init``` and ```resourceWatcherService.add``` is because I'm still not totally confident about the guarantee of no search threads executing while the file is deleted.

By calling ```init()``` first, we know with certainty whether the file existed at the time we started loading the graph, and if it did not exist, we are guaranteed that ```loadIndex``` will fail. If we also start the monitor before loading the index, then in the case that we're wrong about the guarantee mentioned above, we're effectively creating a multi-second window (depending on index loading time) for which the watcher could notify about a file delete, and invalidate() would cause a no-op.

Reply:
makes sense. Thanks